### DISCUSSION TITLE
Add Gradient Accumulation logic in training code to enhance training consistency

### DISCUSSION THREAD

**Discussion by ZetangForward** [OpenRLHF/OpenRLHF#481] [Ideas]
This discussion starts from this PR: https://github.com/OpenRLHF/OpenRLHF/pull/477

## Discussion 1: loss accumulation

I noticed a very important thing: **the loss accumulation logic seems strange in the raw code**:
https://github.com/OpenRLHF/OpenRLHF/blame/5f52e354829e402561313a8c56e0a5c071a3ae65/openrlhf/trainer/sft_trainer.py#L164

In fact, the normal logic for gradient accumulation should 
1. involve the loss being continuously accumulated through `self.strategy.backward`, and then updating the optimizer when `step % self.strategy.accumulated_gradients` is reached; 
2. At the same time, the gradient (loss) for each step should be divided by `accumulated_gradients` rather than directly propagating backward as in the original code

Here's a code snippet that illustrates the correct way to handle gradient accumulation:
```python
# Assuming self.strategy.accumulated_gradients is the number of steps to accumulate gradients
# and self.optimizer is the optimizer instance.

for step, batch in enumerate(train_loader):
    optimizer.zero_grad()
    outputs = model(batch)
    loss = criterion(outputs, targets)
    loss = loss / self.strategy.accumulated_gradients  # Normalize the loss
    self.strategy.backward(loss)  # Accumulate gradients
    if (step + 1) % self.strategy.accumulated_gradients == 0:
        optimizer.step()  # Update the optimizer
        optimizer.zero_grad()  # Reset gradients for the next accumulation cycle
```
This code snippet ensures that the gradients are accumulated over multiple steps and then normalized before updating the optimizer. This approach should help in reducing the discrepancy you observed when using ring attention.

Here is the original code for comparison:
```python
...

loss = ... # CE loss or other loss
self.strategy.backward(loss, self.model, self.optimizer)  # backward the current loss
self.strategy.optimizer_step(self.optimizer, self.model, self.scheduler)  # update the optimizer (no accumulation)

step_bar.update()

loss_mean = loss_mean * 0.9 + 0.1 * gpt_loss.item()
logs_dict = {
    "gpt_loss": gpt_loss.item(),  # here is the mini batch loss not the accumulated loss
    "loss_mean": loss_mean,
    "lr": self.scheduler.get_last_lr()[0],
}

...

# step bar
logs_dict = self.strategy.all_reduce(logs_dict)  # collect all the mini batch loss
step_bar.set_postfix(logs_dict)
step_bar.update()

# logs/checkpoints/evaluation
if step % self.strategy.accumulated_gradient == 0:  # just update the log not the loss
    global_step = step // self.strategy.accumulated_gradient
    client_states = {"consumed_samples": global_step * args.train_batch_size}
    self.save_logs_and_checkpoints(args, global_step, step_bar, logs_dict, client_states)
```

When I tested, I found that the original code would cause a significant deviation when ring attention is turned on or off. 
The reason is that _when ring attention is turned on, there will inevitably be gradient accumulation (because a batch of data is not as much as vanilla attention), but using the original code would result in "fake gradient accumulation."

## Discussion 2: Drop_last of the Dataloader

The data loader currently drops the last samples based on the micro-batch size, not the training batch size. 

https://github.com/OpenRLHF/OpenRLHF/blob/5f52e354829e402561313a8c56e0a5c071a3ae65/openrlhf/cli/train_sft.py#L78

This means that if the total number of samples in an epoch is not perfectly divisible by the micro-batch size, **the remaining samples aren't dropped.  Consequently, these leftover samples might be merged with the new batch at the start of the next epoch.** 

If gradient accumulation is not properly reset at the beginning of each epoch, this could lead to "carryover" effects where the gradients from the previous epoch influence the calculations in the new epoch.

In my modified code, the result is the accumulation of gradients, which is why the vanilla loss and ring attention loss curves in wandb basically overlap (see https://wandb.ai/zecheng-tang/debug_openrlhf_train_sft/workspace?nw=nwuserzechengtang in PR: https://github.com/OpenRLHF/OpenRLHF/pull/477).

I'm not sure if my understanding is correct or if the DeepSpeed within the OpenRLHF library automatically handles the above operations (but upon reviewing the code, it seems like the logic mentioned above is not present). If there are any mistakes, please correct me.

@tsaoyu

**Comment by ZetangForward** [comment]
The modification is quite simple; only two parts need to be changed: one is the data preprocessing, and the other is the training loop.

Below are my implemented code snippets:

## 1. Data Preprocessing

**Current code in main branch:** https://github.com/OpenRLHF/OpenRLHF/blob/eac0e9f122ba398a159d89c579bd9b0d3f092287/openrlhf/utils/utils.py#L41

**My data preprocess logic (def blending_datasets( ) function)**

```python
def blending_datasets(
    datasets,
    probabilities,
    strategy=None,
    seed=42,
    max_count=5000000,
    return_eval=True,
    stopping_strategy="first_exhausted",
    train_split="train",
    eval_split="test",
    train_batch_size=4, # default 4
):
    datasets = datasets.split(",")
    probabilities = list(map(float, probabilities.split(",")))
    assert len(probabilities) == len(datasets)

    train_data_list = []
    eval_data_list = []
    for i, dataset in enumerate(datasets):
        dataset = dataset.strip()
        strategy.print(f"dataset: {dataset}")

        data_dir = dataset.split("@")[1].strip() if "@" in dataset else None
        dataset = dataset.split("@")[0].strip()
        dataset_basename = os.path.basename(dataset)

        ext = os.path.splitext(dataset)[-1]
        # local python script
        if ext == ".py" or (
            os.path.isdir(dataset) and os.path.exists(os.path.join(dataset, f"{dataset_basename}.py"))
        ):
            data = load_dataset(dataset, trust_remote_code=True)
            strategy.print(f"loaded {dataset} with python script")
        # local text file
        elif ext in [".json", ".jsonl", ".csv"]:
            ext = ext.lower().strip(".")
            if ext == "jsonl":
                ext = "json"
            data = load_dataset(ext, data_files=dataset)
            strategy.print(f"loaded {dataset} with data_files={dataset}")
        # local dataset saved with `datasets.Dataset.save_to_disk`
        elif os.path.isdir(dataset):
            data = load_from_disk(dataset)
            strategy.print(f"loaded {dataset} from disk")
        # remote/local folder or common file
        else:
            data = load_dataset(dataset, data_dir=data_dir)
            strategy.print(f"loaded {dataset} from files")

        if train_split and train_split in data:
            train_data = data[train_split].select(range(min(max_count, len(data[train_split]))))
        else:
            train_data = data.select(range(min(max_count, len(data))))
        
        # The residual data causes gradient accumulation issues in the last few steps, 
        # here we automatically discard the extra samples at the end.
        total_size = len(train_data)
        keep_size = (total_size // train_batch_size) * train_batch_size
        train_data = train_data.select(range(keep_size))
        train_data_list.append(train_data)

        if return_eval:
            if eval_split and eval_split in data:
                eval_data = data[eval_split].select(range(min(max_count, len(data[eval_split]))))
            # train will contains eval? TODO
            else:
                eval_data = train_data.select(range(min(max_count, int(len(train_data) * 0.03))))
            eval_data_list.append(eval_data)
```


## 2.  Training loop

**Current code in main branch:** https://github.com/OpenRLHF/OpenRLHF/blob/eac0e9f122ba398a159d89c579bd9b0d3f092287/openrlhf/trainer/sft_trainer.py#L98

**My Training loop (def fit( ) function):**

```python
self.model.train()
loss_mean = 0
accumulated_loss = 0
accumulated_gpt_loss = 0
accumulated_aux_loss = 0

for prompts_id_lens, inputs, attention_masks, packed_seq_lens, infos in self.train_dataloader:
if self.packing_samples:
    inputs = inputs.to(torch.cuda.current_device())
    attention_mask = attention_masks.to(torch.cuda.current_device())
else:
    inputs = inputs.to(torch.cuda.current_device()).squeeze(1)
    attention_mask = attention_masks.to(torch.cuda.current_device()).squeeze(1)

labels = torch.where(attention_mask.bool(), inputs, self.loss_fn.IGNORE_INDEX) # create labels

if not self.pretrain_mode:
    if self.packing_samples:
        index = 0
        for input_length, source_len in zip(infos["input_length"], prompts_id_lens):
            labels[0][index: index + source_len] = self.loss_fn.IGNORE_INDEX
            index += input_length
    else:
        for label, source_len in zip(labels, prompts_id_lens):
            label[:source_len] = self.loss_fn.IGNORE_INDEX

if self.strategy.ring_attn_size == 1: # vanilla attention
    output = self.model(inputs, attention_mask=attention_mask, return_output=True)
    gpt_loss = self.loss_fn(output.logits, labels)

else: # ring attention
    assert self.packing_samples, "Ring attention only works with packing samples"
    num_calculate_tokens = labels.ne(self.loss_fn.IGNORE_INDEX).sum().item()

    local_logits = self.model(
        inputs,
        attention_mask=attention_mask,
        ring_attn_group=self.strategy.ring_attn_group,
        packed_seq_lens=packed_seq_lens, 
        return_output=True,
    )["logits"]
    
    total_seq_len = labels.numel()
    local_seq_len = total_seq_len // self.strategy.ring_attn_size
    
    ########################### loss computation ###########################
    # global labels                     [0,1,2,3,4,5]
    # global masks                      [0,0,0,0,1,1]  # only compute SFT loss on [4, 5]
    # shifted labels                    [1,2,3,4,5,0]
    # local labels (ring_attn_size=2)   [[1,2,3], [4,5,0]]
    ########################### loss computation ###########################
    
    local_slice = slice(rank * local_seq_len + 1, (rank + 1) * local_seq_len + 1)
    
    local_label = labels[:, local_slice]
    if rank == self.strategy.ring_attn_size - 1: # add a dummy label to the last logit
        local_label = F.pad(local_label, (0, 1), value=self.loss_fn.IGNORE_INDEX)
    local_mask = (local_label == self.loss_fn.IGNORE_INDEX)
    
    local_label[local_mask] = 0  # convert -100 in local_label into 0 for `torch.gather` operation
    per_token_logps = torch.gather(local_logits.log_softmax(-1), dim=2, index=local_label.unsqueeze(2)).squeeze(2)

    per_token_logps = per_token_logps * (~local_mask)
    gathered_logps = all_gather(per_token_logps, self.strategy.ring_attn_group)

    gpt_loss = -torch.sum(gathered_logps) / num_calculate_tokens # compute loss on non-masked tokens
    gpt_loss = gpt_loss / self.strategy.accumulated_gradient

# mixtral
if self.aux_loss:
    aux_loss = output.aux_loss
else:
    aux_loss = 0

accumulated_loss += gpt_loss
accumulated_gpt_loss += gpt_loss.item()
if self.aux_loss:
    accumulated_aux_loss += aux_loss.item()

loss = gpt_loss + aux_loss * self.args.aux_loss_coef
self.strategy.backward(loss, self.model, self.optimizer)  # gradient accumulation

step_bar.update()

# logs/checkpoints/evaluation
if step % self.strategy.accumulated_gradient == 0:
    global_step = step // self.strategy.accumulated_gradient
    loss_mean = loss_mean * 0.9 + 0.1 * gpt_loss.item()
    logs_dict = {
        "gpt_loss": gpt_loss.item(),
        "loss_mean": loss_mean,
        "lr": self.scheduler.get_last_lr()[0],
    }
    if self.aux_loss:
        logs_dict["aux_loss"] = accumulated_aux_loss
    logs_dict = self.strategy.all_reduce(logs_dict)
    step_bar.set_postfix(logs_dict)

    self.strategy.optimizer_step(self.optimizer, self.model, self.scheduler) # update model parameters
    client_states = {"consumed_samples": global_step * args.train_batch_size}
    self.save_logs_and_checkpoints(args, global_step, step_bar, logs_dict, client_states)
```

**Comment by ZetangForward** [comment]
Deepspeed can do the above operation
